server/src/main/java/org/elasticsearch/cluster/ClusterSnapshotStats.java

```diff
@@ -0,0 +1,269 @@
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public record ClusterSnapshotStats(
+    int snapshotsInProgressCount,
+    int incompleteShardSnapshotCount,
+    int deletionsInProgressCount,
+    int cleanupsInProgressCount,
+    List<PerRepositoryStats> statsByRepository
+) implements ToXContentObject, Writeable {
+
+    public static ClusterSnapshotStats EMPTY = new ClusterSnapshotStats(0, 0, 0, 0, List.of());
+
+    public static ClusterSnapshotStats of(ClusterState clusterState, long currentTimeMillis) {
+        return of(
+            clusterState.metadata().custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY),
+            clusterState.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY),
+            clusterState.custom(SnapshotDeletionsInProgress.TYPE, SnapshotDeletionsInProgress.EMPTY),
+            clusterState.custom(RepositoryCleanupInProgress.TYPE, RepositoryCleanupInProgress.EMPTY),
+            currentTimeMillis
+        );
+    }
+
+    private static ClusterSnapshotStats of(
+        RepositoriesMetadata repositoriesMetadata,
+        SnapshotsInProgress snapshotsInProgress,
+        SnapshotDeletionsInProgress snapshotDeletionsInProgress,
+        RepositoryCleanupInProgress repositoryCleanupInProgress,
+        long currentTimeMillis
+    ) {
+
+        final var snapshotsInProgressCount = snapshotsInProgress.count();
+        var incompleteShardSnapshotCount = 0;
+        final var deletionsInProgressCount = snapshotDeletionsInProgress.getEntries().size();
+        final var cleanupsInProgressCount = repositoryCleanupInProgress.entries().size();
+        final var perRepositoryStats = new ArrayList<PerRepositoryStats>(repositoriesMetadata.repositories().size());
+
+        for (RepositoryMetadata repository : repositoriesMetadata.repositories()) {
+
+            final var repositoryName = repository.name();
+            final var repositoryType = repository.type();
+
+            var snapshotCount = 0;
+            var cloneCount = 0;
+            var finalizationsCount = 0;
+            var totalShards = 0;
+            var completeShards = 0;
+            final var shardStatesAccumulator = new EnumMap<>(
+                Arrays.stream(SnapshotsInProgress.ShardState.values())
+                    .collect(Collectors.toMap(Function.identity(), ignored -> new AtomicInteger()))
+            );
```

Comment:
NIT: I think it should be possible to use overloaded toMap version that has a map supplier to avoid copying

Reply:
ah yeah so it is